_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,448 @@
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_99.svg" />
+
+As expected, the latency story is an inverse of throughput,
+with 99th percentile being noisiest and showing anomalies in some Rust runs.
+The biggest (geometric) difference is in the 8--16 connections range where Kotlin frameworks already
+saturate available CPU, but Rust does not, yet.
+
+The latency "cutoff" at the very right of the graph corresponds with the change of mode:
+from saturation-induced latency to appearance of error responses.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage.svg" />
+
+Here we measure high-water mark memory usage of the container from its start till the end of each
+benchmark step, as reported by Docker (i.e. not just momentary memory usage).
+
+The comparison between JVM-based and Rust frameworks is not representative here
+as allocated Java heap depends on configuration in addition to actual usage.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage_per_requests.svg" />
+
+The same metric as above, but divided by the number of successful requests per second;
+which gives an unconventional unit of *megabyte-seconds per request*.
+
+It is normal that the values for low connections counts are higher
+because the base memory footprint of each framework is spread out across fewer requests.
+
+You may notice a slight difference in shape between Http4k and the 2 async frameworks:
+Http4k gets down to ~0.05 MiB⋅s and stays there, while the others go up again as latencies increase.
+If you have an explanation for this, please share it in a discussion.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu.svg" />
+
+Consumed CPU time as reported by Docker API.
+
+JVM is seen working hard to JIT-optimise Kotlin-based codebases during the warm-up steps.
+The just-in-time optimisation, combined with the execution of unoptimised code, takes roughly 23 CPU-seconds extra,
+a remarkably small amount of resources.[^jit]
+
+[^jit]: It is tempting to make an unfair comparison to Rust,
+    where optimised build adds 17 CPU-*minutes* on top of 11.5 CPU-minute debug build.
+    We should also acknowledge the artificially low amount of actually executed Kotlin application code in our benchmark.
+
+Still, this is something to keep in mind about JVM-based microservices
+when considering, for example, high-frequency continuous deployment or aggressive instance count autoscaling.
+The JIT may compete for resources with the not-yet-optimised target workload.
+Such characteristic should also apply to other JIT-based engines like Node.js' V8.
+
+Further right in the graph, Kotlin implementations quickly saturate the allocated CPU portion,
+Rust following ~2 steps later.
+
+I'm curious that Http4k manages to consume a bit more CPU than the 15 CPU-seconds assigned by Docker.
```

Comment:
maybe...?

```suggestion
I'm curious how Http4k manages to consume a bit more CPU than the 15 CPU-seconds assigned by Docker.
```

Reply:
Actually, I have an idea *how* this happens: Http4k uses a lot of threads and maybe the CFS scheduler's Bandwidth Control may create a small error for each thread, which accumulates. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/scheduler/sched-bwc.rst#n54 mentions task run-time being synchronized only once per 5ms, but that synchronization between CPU cores, not between threads.

OR, http4k manages to create more *system* (kernel) load than others (again high thread count?), and perhaps Docker limits just userspace load, but includes system load portion in its stats.

Mention that or not? Inline or footnote? I have a hard time deciding whether to include potentially interesting *and* false ideas.